crone/src/BufDiskWorker.cpp

```diff
@@ -128,6 +147,14 @@ int BufDiskWorker::secToFrame(float seconds) {
     return static_cast<int>(seconds * (float) sampleRate);
 }
 
+float BufDiskWorker::raisedCosFade(float unitphase) {
+    return 0.5f * (cosf(M_PI * (1.f + unitphase)) + 1.f);
+}
+
+float BufDiskWorker::mixFade(float x, float y, float a, float b) {
+    return a * x + b * y;
```

Comment:
Split this out to a function so the mixing approach could be tweaked. Not sure if this should use the same mixing calculation as [ReadWriteHead::mixFade](https://github.com/monome/softcut-lib/blob/4b535d9823e59e62437a56fbef6ce382bbcf050f/softcut-lib/src/ReadWriteHead.cpp#L178), I don't think I understand how this works so I left it linear here. Also maybe need to think about clipping behavior?

Reply:
i actually think it's fine to exceed unity range in the buffer.

crossfade options are a whole huge topic that i'd love to discuss...

- for the (generally good[*1]) assumption that real-world signals are uncorrelated, we want the sum of each signal's power (squared amplitude) to remain constant.
- $sin^2 + cos^2 = 1$ is a convenient relation for building curves with that property.
- in crone/softcut we've used both "raised cosine" (for a gradual fade) and "half cosine" for a fast fade.
- tbh, neither is really ideal and don't create "perceptual linearity." [*2]


so anyways maybe we can raise a TODO issue for updating the crossfades perfectly.


[*1] when signals are perfectly correlated, you want "equal voltage" instead of "equal power." it's prohibitive to be determining this in real time for all mix points of course; but you could actually do it for an offline operation.
[*2] it's even weirder in context of softcut xfades, because overdub and record levels are also crossfading, so correlation with signals from the past comes into play.